Update to 1.12.0

andrewdnolan opened the pull request:
## In this version
* Many packages have been updated, including:
  * chemdyg 1.1.1
  * dask 2025.9.1
  * e3sm_diags 3.1.0
  * e3sm_to_cmip 1.13.0
  * esmf 8.9.0
  * esmfpy 8.9.0
  * hdf5 1.14.6
  * libnetcdf 4.9.3
  * livvkit 3.2.0
  * mache 2.0.0
  * matplotlib 3.10.6
  * moab 5.6.0
  * mosaic 1.12.1
  * mpas_analysis 1.14.0
  * mpas_tools 1.4.0
  * mpi4py 4.1.1
  * nco 5.3.6
  * pcmdi_metrics 4.0.1 
  * proj 9.6.2
  * pyproj 3.7.2
  * tempest-extremes 2.4
  * xarray 2025.9.0
  * xcdat 0.10.1
  * zppy 3.1.0
  * zppy-interfaces 0.2.0
  * zstash 1.5.0

* python versions `3.11`, `3.12`, and `3.13` will be supported

## New and improved
- ChemDyg 1.1.0:
    - Updates to be compatible with zppy (v3.0.0)
    - Some minor fixes

- E3SM-Diags 3.0.1 and 3.1.0 :
     - New features:
        - Support for plotting data on native grids (e.g., cubed-sphere, unstructured grids) using UXarray, see [example-native-grid-visualization](https://github.com/E3SM-Project/e3sm_diags/tree/main/examples/ex8-native-grid-visualization)
        - Snapshot analysis on core diagnostic sets (lat_lon, lat_lon_native, polar, zonal_mean_2d, meridional_mean_2d, zonal_mean_2d_stratosphere), allowing analysis of individual time steps of model data, see [example-snapshot-analysis](https://github.com/E3SM-Project/e3sm_diags/tree/main/examples/ex9-snapshot-analysis).  model vs obs support will be included in next releases.
        - Full support of 2d land variable in model vs model run. 
        - Included AR and ETC detection script for standalone use.

    - User-facing Bug fixes and Enhancements
       -  Fixed tropical_subseasonal set by supporting processing year values less than 1000
        - Fixed streamflow diagnostics to support dynamic resolution
        - Added v3LR to baseline in Taylor Diagram and CMIP comparison plots
        - Updated tropical_subseasonal set for EAMxx output

- e3sm_to_cmip v1.13.0 (since 1.11.3):
    - Added support for new v3 atmosphere variable handlers.
    - Introduced 3D aerosol species mixing ratio variable.
    - Centralized results directories, revamped logging, and improved logger readability.
    - Optimized entry point and argument parsing for faster runtime setup.
    - Matched variable handlers by compatible frequency and raw variable names.
    - Fixed MPAS handler logging and temporary directory issues to prevent file collisions and hanging.
    - Ensured unique user_metadata.json files across parallel processes.
    - Added fill_nan() to replace np.nan with proper CMOR fill values.
    - Preserved legacy Xarray compat, join, and data_vars behaviors for stability.
    - View the full [changelog](https://github.com/E3SM-Project/e3sm_to_cmip/releases) for more information.
- LIVVKit 3.1.0:
    - Handles missing performance files more gracefully
    - Improves file diff section header titles
- MOAB 5.6.0:
    - iMOAB, BfB, and remapping: Unified iMOAB API for online and offline linear map workflows, faster parallel SCRIP and domain file reading with dynamic partitioning, correct RLL grid rank propagation, and fully BfB reproducible map projections using arbitrary decompositions.
    - CAAS and visualization: Added parallel CAAS nonlinear bounds-preserving filter and enhanced MOAB VisIt plugin support to visualize linear map couplings alongside meshes and fields.
    - Regression and comparison tools: Expanded regression testing with the compare H5M utility, added new tools to compare projection data and offline maps generated with mbtempest, and improved metadata handling to ensure consistent grid rank/dimension of meshes and fields.
    - Detailed [changelogs](https://bitbucket.org/fathomteam/moab/src/master/RELEASE_NOTES.md) provide more information about all updates.
- mosaic 1.2.1:
    - Mirroring of planar periodic meshes across periodic boundaries. 
    - Better support for directly plotting from restart files. 

- MPAS-Analysis 1.14.0:
    - Enhanced ocean and sea-ice diagnostics: more robust conservation time series, expanded variables for polar regions and mass flux analysis, improved barotropic streamfunction handling, and fixes for hangs and missing projections.
    - Broader and more flexible analysis capabilities: support for custom climatology maps (2D/3D), improved transect processing (including NetCDF transects), addition of wind-stress-curl map plots, and better handling of anomalies and regional time series.

- NCO 5.3.6:
    - `ncremap` fixes OpenMP use for horizontal & vertical regridding
    - `ncclimo` --rgn_stt automatically selects only 2D spatial variables for global timeseries
    - `ncks` S1D ELM gridifier can aggregate fields over all Urban landunit types
    - `ncremap` better detects and warns about corrupt input
- xcdat 0.10.1 (since 0.8.0):
    - Added options to output regridding weights and infer vertical target data.
    - Introduced weight thresholds for spatial and temporal averaging.
    - Improved time frequency inference using median delta calculation.
    - Enhanced coordinate handling and added curvilinear grid support.
    - Fixed major xgcm and xESMF integration issues and regridding alignment.
    - Preserved legacy Xarray open_mfdataset() behaviors to prevent breaking changes.
    - Constrained cf_xarray >= 0.10.7 for correct non-monotonic bounds handling.
    - Removed mask from grid creation to fix xESMF issues.
    - Improved performance and stability across regridding and temporal operations.
    - View the full [changelog](https://xcdat.readthedocs.io/en/latest/history.html) for more information.

- zppy 3.1.0:
    - zppy v3.1.0 adds two new features and is designed for use with zppy-interfaces v0.2.0.
        - Users may now include pcmdi_diags. zppy currently supports mean_climate, variability_modes_cpl, variability_modes_atm, and synthetic_plots. Support for enso will come in a later release. Please note we are considering this to be a "beta release" of pcmdi_diags. Please let the developers know if you encounter any issues by posting on the [discussion page](https://github.com/E3SM-Project/zppy/discussions). Example cfgs using pcmdi_diags are available [here](https://github.com/E3SM-Project/zppy/blob/main/examples/post.v3.LR.amip.0101.cfg#L44) and [here](https://github.com/E3SM-Project/zppy/blob/main/examples/post.v3.LR.historical.0101.cfg#L40).
        - The global_time_series task now has improved performance and clearer distinctions between its operating modes (i.e., a multi-plot PDF or viewers similar to those of e3sm_diags).
    - Other improvements include adding provenance cfg files to the output and web directories for easier debugging, removing machine specific paths, updating the default cfg for ilamb, supporting job_nbr in the ts_task, fixing the regex for tc_analysis, and fixing year looping for land model-vs-model in e3sm_diags.

- zppy-interfaces 0.2.0:
    - Adds support for the new features (pcmdi_diags and improvements to global_time_series) included in zppy v3.1.0.

- zstash 1.5.0:
    - zstash v1.5.0 includes two important updates:
        - Globus authentication has been simplified. Now, instead of running a "toy problem", pasting an auth code twice, and then running the desired command, users can immediately run their desired command and will only be prompted for an auth code once. Additionally, there is no longer a need to delete a token file except when switching to a new destination endpoint.
        - zstash will now detect database corruptions such as a tar appearing multiple times.

## Deployment
Deployed on:
- [x] Andes
- [x] Anvil
- [ ] Aurora
- [x] Chrysalis
- [x] Compy
- [ ] Dane
- [x] Frontier
- [x] Perlmutter-CPU
- [x] ALCF Polaris

## Sync Diagnostics
Synced on:
- [x] Andes
- [x] Anvil
- [x] Aurora
- [x] Chrysalis
- [x] Compy
- [ ] Dane
- [x] Frontier
- [x] Perlmutter-CPU
- [x] ALCF Polaris

## Delete old test envs
- [x] Andes
- [x] Anvil
- [x] Aurora
- [x] Chrysalis
- [x] Compy
- [ ] Dane
- [x] Frontier
- [x] Perlmutter-CPU
- [x] ALCF Polaris

andrewdnolan committed:
Update to v1.12.0rc1

andrewdnolan committed:
Update deployment for mache 1.31.0

andrewdnolan committed:
Update to v1.12.0rc2

Changes default version of python to 3.13 and release versions to
be 3.11, 3.12, and 3.13. To support the new python versions all
CDAT related packages have been removed from the `meta.yaml`
and the associated testing of `cdms2` in the `bootstrap.py` file.
The `pcmdi_metrics` package is currently commented out of the
`meta.yaml` because it has `genutils`, `cdutils`, `cdms2`, and `cdp`
as dependencies. When a new version of `pcmdi_metrics`, without CDAT
dependencies, is released it will be uncommented out of the `meta.yaml`.

With the update to python 3.13 the build string for the python version
has changed and is now dealt with accordingly.

andrewdnolan committed:
Clean up `meta.yaml`

xylar commented:
@andrewdnolan, one thing we need to figure out is how to set permissions on the `base` conda environment and its contents such that multiple maintainers can create environments but users can't install packages into the environment by mistake.

My thought here is that it might be safer to introduce a `conda` directory similar to the `spack` directory and to have a fresh conda installation in each under `e3sm_unified_1_12_0`, etc.  This way, we could set permissions such that `conda` is group writable but `e3sm_unified_1_12_0`, etc. and their contents are only writable by one maintainer.

Could you look into this?  The relevant code to fix would be:
https://github.com/E3SM-Project/e3sm-unified/blob/main/e3sm_supported_machines/shared.py#L132-L133
The function would need to take the version number and convert it to the desired syntax with underscores, similar to what is done for spack path names here:
https://github.com/E3SM-Project/e3sm-unified/blob/main/e3sm_supported_machines/bootstrap.py#L112-L117
But we don't need the machine, compiler or MPI variant because the conda base environment doesn't care about those things.

xylar commented on recipes/e3sm-unified/meta.yaml:
```diff
@@ -1,5 +1,5 @@
 {% set name = "E3SM-Unified" %}
-{% set version = "1.11.1" %}
+{% set version = "1.12.0rc2" %}
```
We decided not to do this.

andrewdnolan committed:
Update TempestExtremes to v2.3.1

A cmake based build system was introduce in version 2.3.0.
The spack package recipe for TempestExtremes uses a cmake build
system, so a version >=2.3.0 is now neeed

andrewdnolan committed:
Move conda installation to versioned directory

Also simplified the spack environment name given that the spack
installation lives within a versioned e3sm-unified direcotry.

xylar commented on recipes/e3sm-unified/meta.yaml:
```diff
@@ -44,18 +44,18 @@ requirements:
+    - mosaic 1.1.0
+    - mpas-analysis 1.14.0rc01
+    - mpas_tools 1.2.2
+    - nco 5.3.4  # [mpi != 'hpc']
```
```suggestion
    - nco 5.3.5  # [mpi != 'hpc']
```
Or maybe newer?

andrewdnolan committed:
Put compiler and mpi in spack directory name

xylar commented on recipes/e3sm-unified/meta.yaml:
```diff
@@ -44,18 +44,18 @@ requirements:
-    - nco 5.3.2  # [mpi != 'hpc']
-    - pcmdi_metrics 3.8.2
+    - mosaic 1.1.0
+    - mpas-analysis 1.14.0rc01
```
```suggestion
    - mpas-analysis 1.14.0rc2
```
https://github.com/conda-forge/mpas-analysis-feedstock/pull/75

xylar commented on recipes/e3sm-unified/meta.yaml:
```diff
@@ -1,5 +1,5 @@
 {% set name = "E3SM-Unified" %}
-{% set version = "1.11.1" %}
+{% set version = "1.12.0rc2" %}
```
Please also remove:
```
    - nb_conda
    - nb_conda_kernels
```

xylar replied:
We want e3sm_to_cmip 1.13.0rc1:
https://github.com/conda-forge/e3sm_to_cmip-feedstock/pull/42

xylar commented on recipes/e3sm-unified/meta.yaml:
```diff
@@ -44,18 +44,18 @@ requirements:
-    - pcmdi_metrics 3.8.2
+    - mosaic 1.1.0
+    - mpas-analysis 1.14.0rc01
+    - mpas_tools 1.2.2
```
```suggestion
    - mpas_tools 1.4.0rc1
```
https://github.com/conda-forge/mpas_tools-feedstock/pull/138

xylar commented on e3sm_supported_machines/default.cfg:
```diff
@@ -39,13 +39,13 @@ xesmf = None
+hdf5 = hdf5@1.14.6+cxx+fortran+hl+mpi+shared
 moab = moab@5.5.1+mpi+hdf5+netcdf+pnetcdf+metis+parmetis+tempest
-nco = nco@5.3.2+openmp
+nco = nco@5.3.4+openmp
```
```suggestion
nco = nco@5.3.5+openmp
```

xylar commented on e3sm_supported_machines/default.cfg:
```diff
@@ -39,13 +39,13 @@ xesmf = None
 parallel_netcdf = parallel-netcdf@1.12.3
 # parallelio = parallelio@2.5.10+fortran+mpi~pnetcdf
-tempestextremes = tempestextremes@2.2.3+mpi
+tempestextremes = tempestextremes@2.3.1+mpi
```
I created the branch:
https://github.com/E3SM-Project/spack/tree/spack_for_mache_2.0.0
It has the new tempestextremes and nco (above).  For the former, there could, of course, be bugs since I just copied the CMakePackage from spack-packages.

andrewdnolan committed:
Apply suggestions from code review

Update package version to be in sync with confluence page

Co-authored-by: Xylar Asay-Davis <xylarstorm@gmail.com>

andrewdnolan commented on recipes/e3sm-unified/meta.yaml:
```diff
@@ -75,24 +75,17 @@ requirements:
-    - hdf5 1.14.3 {{ mpi_prefix }}_*
+    - hdf5 1.14.6 {{ mpi_prefix }}_*
     - ipygany
     - libnetcdf 4.9.2 {{ mpi_prefix }}_*
```
Didn't add this to the batch commit, because I wanted to make sure I updated the `netcdf_c` entry in `e3sm_supported_machines/default.cfg ` at the same time.

andrewdnolan committed:
Manually apply more suggestions from code review

andrewdnolan committed:
Add zppy-interfaces dev label

andrewdnolan committed:
Explicitly add notebook as dependecy

Remove constraint on notebook which was for compatability
with nb_conda, which has been removed. notebook has been
left in, without constraint, so that jupyter notebooks are
still avaialable

andrewdnolan committed:
Remove testing command for nb_conda

andrewdnolan committed:
Sync meta.yaml with confluence page

andrewdnolan committed:
If depolying a RC add the dev label to channels

xylar commented:
## Group write permission

I enabled group write permission on the E3SM-Unified base directories on:
- [x] Andes
- [x] Anvil
- [x] Aurora
- [x] Chrysalis
- [ ] Compy
- [ ] Dane
- [x] Frontier
- [x] Perlmutter-CPU
- [ ] ALCF Polaris
- [ ] Ruby

I don't have access to Ruby or Dane and I wasn't able to log on to Polaris.  I am renewing my Compy access.

andrewdnolan committed:
fix return variables name

andrewdnolan committed:
Bump mpi4py version

xylar commented on recipes/e3sm-unified/meta.yaml:
```diff
@@ -42,22 +42,22 @@ requirements:
-    - mache 1.31.0
+    - livvkit 3.2.0
+    - mache 1.32.0
     - moab 5.5.1 {{ mpi_prefix }}_tempest_*  # [mpi != 'hpc']
```
We merged https://github.com/conda-forge/moab-feedstock/pull/106 and this is no longer needed.  Make sure to remove it in the next RC.

xylar committed:
Add "Verify compute-node activation" step to deployment docs

andrewdnolan committed:
uncomment file permsission updates

andrewdnolan committed:
Update to v1.12.0rc2

xylar committed:
Shorten e3sm-unified deployment path

This is necessary to avoid a SbangPathError in spack on Frontier.

This merge also drops the requirement that the machine be specified.
There is no longer an attempt to use a shared conda base environment
as the conda environment for bootstrapping.  Instead, the person
deploying Unified must supply the `--conda` flag or have a base
environment loaded before deploying.

xylar committed:
Clean up get_conda_base()

The deploy and bootstrap scripts are now getting totally different
conda bases in totally different ways so a shared function doesn't
make sense anymore.  This merge splits the functionality back into
these 2 scripts for better clarity and simplicity.

andrewdnolan commented on e3sm_supported_machines/shared.py:
```diff
@@ -127,14 +128,14 @@ def install_miniforge3(conda_base, activate_base):
-            conda_base = os.path.abspath(os.path.join(
-                config.get('e3sm_unified', 'base_path'), 'base'))
+            base_path = config.get('e3sm_unified', 'base_path')
+            conda_base = os.path.join(base_path, env_name, machine, 'conda')
```
Agreed, I think this is totally the right way forward.

andrewdnolan committed:
Update paths the need permission changes

Now consitent with the new layout for e3sm-unified

andrewdnolan committed:
Start user restricted perms at machine dir level

xylar committed:
Add `pre_conda_script` to load scripts

This is needed to unload a problem module on ALCF Polaris and
may be useful on other machines in the future.

This requires a new capability in mache.

andrewdnolan committed:
Update symlinks created for NCO

andrewdnolan committed:
Update to v1.12.0rc3

andrewdnolan committed:
Update version number in meta.yaml

andrewdnolan committed:
Add check for ${PBS_JOBID} to templates

andrewdnolan committed:
Ensure prebuilt wheel for mpi4py is not installed

andrewdnolan committed:
Update base path permission non recursively

andrewdnolan committed:
Update recipes/e3sm-unified/meta.yaml

Co-authored-by: Xylar Asay-Davis <xylarstorm@gmail.com>

andrewdnolan committed:
Update to 1.12.0rc4

Adds moab and mache RCs

andrewdnolan committed:
Pass "include_e3sm_hdf5_netcdf" as kwarg Fixes #140

xylar commented on e3sm_supported_machines/default.cfg:
```diff
@@ -39,13 +39,13 @@ xesmf = None
-netcdf_c = netcdf-c@4.9.2+mpi~parallel-netcdf
+esmf = esmf@8.9.0+mpi+netcdf~pnetcdf~external-parallelio
+hdf5 = hdf5@1.14.6+cxx+fortran+hl+mpi+shared
+moab = moab@5.6.0rc2+mpi+hdf5+netcdf+pnetcdf+metis+parmetis+tempest
```
@vijaysm, out of curiosity, why `~shared`?  I would think shared libraries would be safe and preferable in E3SM-Unified on compute nodes.

andrewdnolan replied:
Sorry about the slow response, but I like @vijaysm idea of keeping this consistent throughout unified.

andrewdnolan committed:
Upate to 1.12.0

xylar reviewed (approved):
Looks good.  Let's get this merged.
